### PR TITLE
Extension for Rational Bezier Splines

### PR DESCRIPTION
**Pull request by jzwar**
# Extension for Rational Bezier Splines
| :warning: WARNING          |
|:---------------------------|
| This PR is blocked from PR   [Rational Bezier Splines in Splinepy](https://github.com/tataratat/splinepy/pull/20) |
## Related Issues
#10 relates to new implementation of rational splines from within bezman
## Description
This PR extends gustaf's splines with rational Bezier splines. It further facilitates the conversion between splines, as from now on Bezier type splines can be converted into their non-uniform counterparts from `SplineLib`. These types can similarly be converted into a list of patches of the underlying part using
```
<BSpline-object>.extract.beziers()
or
<NURBS-object>.extract.beziers()
```
For type-safety, the `extract_bezier_patches()` routine was overwritten in the gustaf spline implementation.

Other changes:
-`GustavBase` and `GustavSpline` were changed into `GustafBase`, `GustafSpline`, respectively
- Updated Documentation of Bezier Splines were some copy-paste errors occured

## Remaining Work
- Extend Export / IO routines from Bezman

### CODE REVIEW

File: gustaf/spline/base.py [hunk 1]
```diff
@@ -440,8 +440,103 @@ def __init__(
...
+        --------
+        same_bspline : BSpline
+        """
+        return BSpline(
+            degrees=self.degrees,
+            control_points=self.control_points,
+            knot_vectors=[
+              [0] * (self.degrees[i] + 1) + [1] * (self.degrees[i] + 1)
+              for i in range(self.para_dim)]
+          )
```
Comment (j042) [inline gustaf/spline/base.py:462]:
```suggestion
            knot_vectors=[[0] * dp1 + [1] * dp1 for dp1 in self.degrees + 1],
          )
```

File: gustaf/spline/base.py [hunk 2]
```diff
@@ -440,8 +440,103 @@ def __init__(
...
+            degrees=self.degrees,
+            control_points=self.control_points,
+            knot_vectors=[
+              [0] * (self.degrees[i] + 1) + [1] * (self.degrees[i] + 1)
+              for i in range(self.para_dim)],
+            weights=self.weights
+          )
+
+
+class BSpline(GustafSpline, splinepy.BSpline):
```
Comment (j042) [inline gustaf/spline/base.py:539]: `gustav` worked hard

File: gustaf/spline/base.py [hunk 3]
```diff
@@ -450,7 +545,7 @@ def __init__(
             control_points=None,
     ):
         """
-        BSpline of gustaf. Inherited from splinepy.BSpline and GustavSpline.
+        BSpline of gustaf. Inherited from splinepy.BSpline and GustafSpline.
```
Comment (j042) [inline gustaf/spline/base.py:548]: `gustaf` serves you

File: gustaf/spline/base.py [hunk 4]
```diff
@@ -564,6 +678,25 @@ def _mfem_ids(self):
...
+      ----------
+      None
+
+      Returns
+      -------
+      None
+      """
+      logging.warning(
+          "Functionality not supported, please use:\n"
+          "<Nurbs>.extract.beziers()"
```
Comment (j042) [inline gustaf/spline/base.py:696]:
```suggestion
          "<NURBS>.extract.beziers()"
```

File: gustaf/spline/base.py [hunk 5]
```diff
@@ -499,8 +594,27 @@ def nurbs(self):
             weights=np.ones(self.control_points.shape[0], dtype="float64")
         )
 
+    def extract_bezier_patches(self):
+      """
+      Overwrites splinepy-parent's function to ensure the conversion of Splines
+      into a usable gustaf format
```
Comment (j042) [inline gustaf/spline/base.py:600]: nice

File: gustaf/spline/base.py [hunk 6]
```diff
@@ -499,8 +594,27 @@ def nurbs(self):
...
+      """
+      logging.warning(
+          "Functionality not supported, please use:\n"
+          "<BSpline>.extract.beziers()"
+      )
+      return None
 
-class NURBS(GustavSpline, splinepy.NURBS):
+
+class NURBS(GustafSpline, splinepy.NURBS):
```
Comment (j042) [inline gustaf/spline/base.py:617]: `gustav` is OG

File: gustaf/spline/extract.py [hunk 7]
```diff
@@ -395,6 +395,26 @@ def control_mesh(spline,):
...
+def beziers(spline):
+  """
+  Extracts Bezier-type objects of any spline-type object
+
+  Parameters
+  ----------
+  spline : Gustaf-Spline
+    """
+  from gustaf.spline.base import Bezier, RationalBezier
+  if "Bezier" in spline.whatami:
```
Comment (j042) [inline gustaf/spline/extract.py:407]:
```suggestion
  from gustaf.spline.base import Bezier, RationalBezier
  
  if "Bezier" in spline.whatami:
```

File: gustaf/spline/extract.py [hunk 8]
```diff
@@ -395,6 +395,26 @@ def control_mesh(spline,):
...
+      return spline
+  elif "BSpline" in spline.whatami:
+      return [Bezier(**s.todict()) 
+          for s in super(type(spline), spline).extract_bezier_patches()]
+  elif "NURBS" in spline.whatami:
+      return [RationalBezier(**s.todict()) 
+          for s in super(type(spline), spline).extract_bezier_patches()]
+  else:
+      raise ValueError("Unknown Spline-Type.")
```
Comment (j042) [inline gustaf/spline/extract.py:416]:
```suggestion
      raise TypeError("Unknown Spline-Type.")
```

File: gustaf/spline/extract.py [hunk 9]
```diff
@@ -395,6 +395,26 @@ def control_mesh(spline,):
...
+  Parameters
+  ----------
+  spline : Gustaf-Spline
+    """
+  from gustaf.spline.base import Bezier, RationalBezier
+  if "Bezier" in spline.whatami:
+      return spline
+  elif "BSpline" in spline.whatami:
+      return [Bezier(**s.todict()) 
+          for s in super(type(spline), spline).extract_bezier_patches()]
```
Comment (j042) [inline gustaf/spline/extract.py:411]:
```suggestion
      return [
          Bezier(**s.todict()) 
          for s in super(type(spline), spline).extract_bezier_patches()
      ]
```

File: gustaf/spline/extract.py [hunk 10]
```diff
@@ -395,6 +395,26 @@ def control_mesh(spline,):
...
+  elif "NURBS" in spline.whatami:
+      return [RationalBezier(**s.todict()) 
+          for s in super(type(spline), spline).extract_bezier_patches()]
```
Comment (j042) [inline gustaf/spline/extract.py:414]:
```suggestion
      return [
          RationalBezier(**s.todict()) 
          for s in super(type(spline), spline).extract_bezier_patches()
      ]
```
Generally, we should avoid multi-line list-comprehensions. But they are just too attractive I guess

File: gustaf/vertices.py [hunk 11]
```diff
@@ -10,9 +10,9 @@
 from gustaf import settings
 from gustaf import utils
 from gustaf import show
-from gustaf._base import GustavBase
+from gustaf._base import GustafBase
```
Comment (j042) [inline gustaf/vertices.py:13]: `gustaf` is welcomed

File: gustaf/vertices.py [hunk 12]
```diff
@@ -10,9 +10,9 @@
...
 
-class Vertices(GustavBase):
+class Vertices(GustafBase):
```
Comment (j042) [inline gustaf/vertices.py:15]: `gustav`, you were awesome. You are awesome. You were the best geometry butler I've ever had. Thank you so much for all the amazing work you've done. I am happy you get to retire finally! I will be visiting you. We will be visiting you. Look #8, #15, it is already happening! Call me, I will take you to RAM whenever you want.  Love, J

**Review by j042 — CHANGES_REQUESTED** [review]
Amazing, good stuff!

File: gustaf/spline/base.py [hunk 13]
```diff
@@ -425,7 +426,6 @@ def __init__(
         Parameters
         -----------
         degrees: (para_dim,) list-like
-        knot_vectors: (para_dim, ...) list
```
[the remark this answers is not in the document]
Comment (jzwar) [reply gustaf/spline/base.py:428]: I don't know what you mean

File: gustaf/vertices.py (hunk 12 above)
Comment (jzwar) [reply gustaf/vertices.py:15]: 👋

**Review by j042 — APPROVED** [review]
Exquisite. Thanks!

**Pull request merged**